DOCSITE: Reorganise tutorials and add inline solutions

bbrcknl opened the pull request:
**Aim:** Arrange the tutorial material such that it is clear and easy to follow
**Solution:** Inspired by the Rust book https://doc.rust-lang.org/book/, the user can see the chapters in the index on the left, and go straight to the sections that they need

Specifically, updated tutorials with: 
  - a streamlined guide to completing the tutorials
  - a streamlined "setting up your machine" page
  - a how-to page with solutions to tutorial questions
  - more organised landing page

To-dos:
 - The new tutorial md files *are not* those from the tutorials repo. The md files in the tutorials repo need to be updated to match the tutorials in this PR, which include solutions and corrections. The new tutorial md files then need to be rewritten as includes to read in the md files from the tutorials repo.
 - Sections on the docsite, other than the tutorials, need to be streamlined

Indanz commented:
In the preview you gave it looked great, good work!

> To-dos:
> 
>     * The new tutorial md files _are not_ those from the tutorials repo. The md files in the tutorials repo need to be updated to match the tutorials in this PR, which include solutions and corrections. The new tutorial md files then need to be rewritten as includes to read in the md files from the tutorials repo.

Because this hasn't been done yet, I can't see what your changes to the tutorial files themselves are, making reviewing it difficult. I can do a manual diffs between _repos/sel4proj/sel4-tutorials/tutorials/* and Tutorials/*, but that is cumbersome.

I used the following diff to build and preview the site locally:

```diff
diff --git a/Gemfile b/Gemfile
index 9e06a50e54..363f7f04eb 100644
--- a/Gemfile
+++ b/Gemfile
@@ -27,6 +27,7 @@ group :jekyll_plugins do
   gem 'jekyll-toc'
   gem 'jekyll-redirect-from'
   gem 'jekyll-remote-theme'
+  gem "jekyll-sass-converter", "~> 2.0" # TODO: Prevent 3.0 untill sass warnings are fixed
 end
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
diff --git a/_config.yml b/_config.yml
index 65f4b8b9a2..b56c93e493 100644
--- a/_config.yml
+++ b/_config.yml
@@ -129,3 +129,4 @@ exclude:
   - assets/css/highlighting/demos
   - assets/css/highlighting/tools
   - assets/css/highlighting/index.html
+  - Tutorials-old
```

On the http://0.0.0.0:4000/Tutorials/ page, the two blocks on the right don't seem right and point to the wrong url? Maybe they shouldn't always be generated, but only on some pages where it makes sense.

The "Tutorials about seL4" page has some broken urls (e.g. "About seL4"), mostly missing the new subdirectories you added for the tutorials, but other pages also have broken links.

The "Quick solution" need a bit more margin below them and it would help if the mouse pointer would change when hovering above them, both can be fixed with a small css change. Also, when adding a new CSS rule, remove all existing styles from the HTML.

It would be nice if, when clicking on a link in "How to: A quick solutions guide", the Quick solution would be expanded automatically. We can add this in the future, but it will require some Javascript (It needs to check a GET or POST variable and set `open` to `true` for all `details` [elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details)).

All in all my first impression is that the new structure is a lot better and makes more sense compared to what we have now. I'll give more detailed feedback later, maybe when the changes are moved to the tutorials repo.

bbrcknl commented:
@Indanz 
> In the preview you gave it looked great, good work!

Thanks!

> > To-dos:
> > ```
> > * The new tutorial md files _are not_ those from the tutorials repo. The md files in the tutorials repo need to be updated to match the tutorials in this PR, which include solutions and corrections. The new tutorial md files then need to be rewritten as includes to read in the md files from the tutorials repo.
> > ```
> 
> Because this hasn't been done yet, I can't see what your changes to the tutorial files themselves are, making reviewing it difficult. I can do a manual diffs between _repos/sel4proj/sel4-tutorials/tutorials/* and Tutorials/*, but that is cumbersome.

I suggest waiting to review the tutorials until I've made the changes in the tutorials repo, then you'll be able to see line-by-line diffs.
I will start that early next week and keep you posted. It will take ~2 weeks.

> On the http://0.0.0.0:4000/Tutorials/ page, the two blocks on the right don't seem right and point to the wrong url? Maybe they shouldn't always be generated, but only on some pages where it makes sense.

I have added two new layouts to the Tutorial pages to remove the blocks that don't make sense.
_layouts/overview.html
_layouts/tutorial.html


> The "Tutorials about seL4" page has some broken urls (e.g. "About seL4"), mostly missing the new subdirectories you added for the tutorials, but other pages also have broken links.
> 
The urls have been fixed.

> The "Quick solution" need a bit more margin below them and it would help if the mouse pointer would change when hovering above them, both can be fixed with a small css change. Also, when adding a new CSS rule, remove all existing styles from the HTML.
> 
TBD: Integrate CSS with markdown for quick solutions

> It would be nice if, when clicking on a link in "How to: A quick solutions guide", the Quick solution would be expanded automatically. We can add this in the future, but it will require some Javascript (It needs to check a GET or POST variable and set `open` to `true` for all `details` [elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details)).
> 

TBD: Integrate js with markdown to expand solutions automatically

> All in all my first impression is that the new structure is a lot better and makes more sense compared to what we have now. I'll give more detailed feedback later, maybe when the changes are moved to the tutorials repo.

bbrcknl commented:
The tutorials have now been updated to fit into the new tutorial structure as per this PR (#231).

The new tutorials: https://github.com/seL4/sel4-tutorials/pull/106

A note on using the rewritten docsite (#231):
While the docsite in this PR has the new tutorial layout, the **Makefile clones the old tutorials** from https://github.com/seL4/sel4-tutorials.
You’ll **need to integrate the updated tutorials** (from [this PR](https://github.com/seL4/sel4-tutorials/pull/106)) in the Makefile from https://github.com/seL4/docs/pull/231

Previous TODOs:

> The "Quick solution" need a bit more margin below them and it would help if the mouse pointer would change when hovering above them, both can be fixed with a small css change. Also, when adding a new CSS rule, remove all existing styles from the HTML.

- I’ve added a margin below the “Quick solutions”
- I haven’t updated the mouse to hover over the solution.
- @Indanz What do you mean by “when adding a new CSS rule, remove all existing styles from the HTML”?

> It would be nice if, when clicking on a link in "How to: A quick solutions guide", the Quick solution would be expanded automatically. We can add this in the future, but it will require some Javascript (It needs to check a GET or POST variable and set open to true for all details [elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details)).

- I’ve integrated js with markdown to expand solutions automatically, but only when coming from the How-to page. Is this sufficient, or do we also need an “Expand all solutions” button on each tutorial page?

lsf37 commented:
(rebased to more easily see the new commits)

lsf37 commented on .ruby-version:
```diff
@@ -1 +1 @@
-2.7.8
+2.7.7
```
Why downgrade the Ruby version? Is there something in 2.7.8 that breaks in the new setting?

bbrcknl replied:
Using 2.7.8, I get an error when running `bundle update`.
```
LoadError: incompatible library version - /Users/jblogs/docs/vendor/bundle/ruby/2.7.0/gems/strscan-3.1.0/lib/strscan.bundle
```
It may be an Apple Silicon issue but I don't know enough about it to diagnose and fix.

lsf37 replied:
Right, ok. There are two issues, I think:

- you shouldn't need to run `bundle update` -- that will try to get new versions of everything and update the lock file, which we do sometimes do, but only to explicitly update dependencies
- the `Makefile` will run `bundle install` which should install what is in `Gemfile.lock`, it's Ok to run that one manually as well
- `bundle install` may fail if you already have gems/libraries installed in the `vendor` directory from another version (either ruby version or from something new that `bundle update` brought in). My first try if you get that error message above would be to delete the `vendor` directory and retry `bundle install`.

I'll have a go at checking if that works on M1 and intel on the branch.

bbrcknl replied:
I deleted the ```vendor``` directory and now I can run 2.7.8. Thanks!

lsf37 replied:
Cool! I'll drop the commits for the ruby version then.

lsf37 commented on Gemfile.lock:
```diff
@@ -127,4 +127,4 @@ DEPENDENCIES
 
 BUNDLED WITH
-   2.1.4
+   2.3.22
```
This doesn't quite fit with the ruby version. Probably just needs to be left out.

bbrcknl replied:
This output is automated with `bundle update`. Now I get `BUNDLED WITH  2.4.22` when I run `bundle update` (still using Ruby 2.7.7).

lsf37 commented:
Ok, I've managed to get it working, and before I start reviewing all the small details, it's probably better to start high level:

- The new inline solutions and the resulting how-to guide are really cool!
- I still need to go through all the parts, but I also like the new flow. There are some things to update in the details, though:
  - the "about seL4 page" in the beginning doesn't really fit. It's too long-winded and not aimed at tutorial level. We should probably start with the overview page instead (also not the Microkit).
  - the Microkit page calls the Microkit "an operating system framework". I don't think that name is right, and I'm not sure the page as such helps much. We might want to replace links to this page directly with links to the Microkit tutorial (not sure if that always works).
  -  the overview page says "a basic understanding of seL4, e.g. by having done the Microkit tutorial." -- the Microkit tutorial is mostly hiding seL4 and not really providing much understanding on what seL4 is or does. The seL4 white paper might be a better resource for this one. I'm fine with pointing people at the Microkit tutorial as a potential very first step, but it needs to be clear that you won't really learn much about seL4 that way -- you'll learn how to build systems with a static architecture on top of seL4 there, which might be all that people want, but it's slightly different to the rest.
  - the overview page doesn't really give an overview of the tutorials. It should. It should explain when you want to do/use Microkit, seL4 directly and/or CAmkES. Also what MCS is, what the dynamic libraries are about, and that there is Rust support.
  - the prerequisites look a bit scary -- it's good that they are there, but it looks like I first have to read all of these before I can run "Hello world", which is not the case. You do need a background in OS for many of the other things, so I don't think we should remove the section, but we might for instance kick out the two wikipedia links that aren't really much good for learning anything (and if you do read the OS book, you'll get enough background for these).
 
- the new front page doesn't really work for me -- possibly because I have the browser on half screen size by default (which is similar to tablet size), and having 3 columns there looks not so great. The old 2 column design is about as much as can fit with that column size. How about leaving the main design intact for now, just changing the links under "Tutorials" there? (and making "Tutorials" itself not a link). Could still be tutorials, howtos, and manual. I would possibly add API Reference if there is space
- a lot of the general menu links in unrelated parts of the site are now rendered as going off-site (probably just some small tweak somewhere, but needs to be fixed)
- I'd replace the links to forum, etc in the Resources section of the nav section on the left with a single link to a contact page
- Not sure the rename from "Getting Started" into "Working with seL4" is a success. The problem is more the content of that page, not so much what it is called ;-). In any case, "Working with seL4" is too long -- it line-breaks on narrower screens and looks bad when it does. Whatever name we pick should be narrower than the min size of the rest of the card on the front page. It'd be great to have an actual good Getting Started page, but that might lead a bit far astray from the purpose of this PR. I'd be fine with removing the links to that page from the top-level navigation and calling the heading on the front page "Resources" or something generic like that (without link to this specific page).

lsf37 commented on Makefile:
```diff
@@ -51,7 +51,7 @@ _repos/tutes:
 	PYTHONPATH=_repos/sel4/capdl/python-capdl-tool _repos/sel4proj/sel4-tutorials/template.py --docsite --out-dir _repos/tutes --tut-file $</$(@F)
 
-TUTORIALS:= $(filter-out index.md,$(notdir $(wildcard Tutorials/*.md)))
+TUTORIALS:= $(filter-out index.md overview.md setting-up.md get-the-tutorials.md,$(notdir $(wildcard Tutorials/seL4Kernel/*.md Tutorials/CAmkES/*.md Tutorials/DynamicLibraries/*.md Tutorials/MCS/*.md)))
```
Something about the path `seL4Kernel` bugs me :-). It should be either `seL4` or `kernel`, but not `seL4Kernel`.

lsf37 replied:
Slightly more content-oriented -- should we really make a new directory structure here? Would it make sense to keep the directory names of the `sel4-tutorials` repo, or does the structure here get translated directly into navigation structure on the site?

bbrcknl replied:
@lsf37 

> Something about the path `seL4Kernel` bugs me :-). It should be either `seL4` or `kernel`, but not `seL4Kernel`.

What about `mechanisms` since these are the `mechanisms` tutorials?

lsf37 replied:
`mechanisms` doesn't really say what the mechanisms apply to. Is `seL4` not good?

bbrcknl replied:
I got `mechanisms` from the current tutorials overview page https://docs.sel4.systems/Tutorials/, from the **seL4 mechanisms tutorials** section.

But yeah, `seL4` is more specific.

bbrcknl replied:
> Slightly more content-oriented -- should we really make a new directory structure here? Would it make sense to keep the directory names of the `sel4-tutorials` repo, or does the structure here get translated directly into navigation structure on the site?

Fair point. I'm working on reverting to 
```
TUTORIALS:= $(filter-out index.md,$(notdir $(wildcard Tutorials/*.md)))
```
but it's taking a while to fix all the broken links.

lsf37 commented on Tutorials/CAmkES/camkes-vm-linux.md:
```diff
@@ -3,9 +3,12 @@ toc: true
 SPDX-FileCopyrightText: 2020 seL4 Project a Series of LF Projects, LLC.
 ---
 {% include tutorial.md %}
+<script src="{{ base.url | prepend: site.url }}/assets/js/toggle-markdown.js"></script>
```
If possible we should move `script` tags and similar into the main layout template (I think `_layouts/tutorial.html` in this case)

lsf37 commented on Tutorials/index.md:
```diff
@@ -1,253 +1,32 @@
 We have developed a series of tutorials to introduce seL4 and
 developing systems on seL4.
 
-## How to use the tutorials
-
-Depending on your goals and what you want to do with seL4, we suggest
-different paths to follow through the tutorial material.  Choose the
-most relevant category below and follow the tutorials in the suggested
-order.
```
I do think that section and set of paths was useful. We should not remove that completely, just recommend a default path (which the site navigation basically suggests now). There is a lot of useful content overview in here that we could use in the overview page.

bbrcknl replied:
Much of this page has been put into the new Overview and Pathways pages, which describe the tutorials and alternate pathways respectively.

lsf37 commented on Tutorials/index.md:
```diff
@@ -1,253 +1,32 @@
-of [libraries and
-components](/projects/available-user-components.html) available to
-build seL4-based systems.
+<h2>Categories</h2>
+The tutorials are split into a number of broad categories:
+
+- [About seL4](GettingStarted/about-seL4) and [Getting started with the Microkit tutorial](GettingStarted/microkit) introduce seL4 concepts.
+- The [seL4 Kernel tutorials](seL4Kernel/overview.md) are a deep dive into seL4 concepts.
+- [MCS](MCS/mcs) introduces seL4 MCS extensions.
+- [Dynamic Libraries](DynamicLibraries/dynamic-1) covers the libraries that have been developed for rapidly prototyping systems on seL4.
+- [CAmkES](CAmkES/hello-camkes-0) is a platform for building componentised systems for embedded platforms.
+- [Microkit](https://trustworthy.systems/projects/microkit/tutorial/)is an operating system framework on top of seL4 provides a small set of simple abstractions that ease the design and implementation of statically structured systems on seL4. (Links to the same tutorial as in the [Getting Started](GettingStarted/microkit) section.)
+- [Rust](https://github.com/seL4/rust-sel4) provide crates for supporting the use of Rust in seL4 userspace.
```
I did not come across this content when I was navigating the new site -- which page does it get rendered to?

bbrcknl replied:
This page was rendered when the user clicked on "Tutorials." I've now redirected "Tutorials" to overview, and used much of the above content in the overview page.

lsf37 commented on Tutorials/seL4Kernel/get-the-tutorials.md:
````diff
@@ -0,0 +1,32 @@
+## Python Dependencies
+Additional python dependencies are required to build [tutorials](ReworkedTutorials). To install you can run:
+```
+pip install --user aenum
+pip install --user pyelftools
````
```suggestion
pip3 install --user aenum pyelftools
```
 
These packages should already be included in `camkes-deps` which you'd need at least when you get to the CAmkES parts, but possibly already before. W should make sure that the standard install instructions tell people to do `pip3 install --user camkes-deps`, which should get them everything that is needed.

bbrcknl replied:
Is it sufficient (and correct) to say

# Getting the tutorials
## Python Dependencies
Additional python and CAmkES dependencies are required to build [tutorials](ReworkedTutorials). To install you can run:
```
pip3 install --user aenum pyelftools camkes-deps

lsf37 replied:
Have put a new comment on the new file so it's easier to track. We only need `camkes-deps`, the `aenum` and `pyelftools` are already contained in that.

lsf37 commented on Tutorials/seL4Kernel/setting-up.md:
```diff
@@ -0,0 +1,237 @@
+---
+
+# Setting up your machine
+**Overview**
+- Set up your machine - install dependencies required to run seL4``
+- Run seL4test on a simulator
+- Gain awareness of terminology used for seL4
```
These should either be links or a paragraph. "Overview" should be a heading.

bbrcknl replied:
I've removed this, since the contents list on the right provides links to each of the sections that are on the page.

lsf37 commented on Tutorials/seL4Kernel/setting-up.md:
```diff
@@ -0,0 +1,237 @@
+
+
+
+## Docker
+To compile and use seL4, it is recommended that you use Docker to isolate the dependencies from your machine.
```
If we are recommending docker, we don't really need the rest of the page apart from `repo` (although I don't know if the virtual machine tutorials actually work under docker). We either should remove the docker section or add some text that makes clear that docker and the rest are two alternatives, and that you need only one of them.

bbrcknl replied:
I actually removed everything but the repo and docker sections, because there is a whole other page on dependencies https://docs.sel4.systems/projects/buildsystem/host-dependencies.html, should someone need them.

I couldn't get of the virtual machine tutorials working on my silicon mac (but I don't think it's related to Docker at all).

Adding issues here for completeness:

"CAmkES VM: Adding a Linux Guest" uses KVM and I would have needed an alternative for mac. 

"CAmkES VM: Cross VM Connectors" breaks using `ninja`.

```
sh.ErrorReturnCode_1: 

  RAN: /usr/bin/cmake -G Ninja -DTUT_BOARD=pc -DTUT_ARCH=x86_64 -DFORCE_IOMMU=ON -DTUTORIAL_DIR=camkes-vm-crossvmke7jqji2 -C ../projec

  STDOUT:


  STDERR:
```

Should I add a note to these tutorials saying the instructions are for Linux only?

lsf37 replied:
Yes, that would be good. We've already had some questions from people who were trying to use other platforms.

lsf37 commented on Tutorials/seL4Kernel/setting-up.md:
```diff
@@ -0,0 +1,237 @@
+
+## Google's Repo tool
+
+The primary way of obtaining and managing seL4 project source is through the use of Google's Repo tool.
```
```suggestion
The primary way of obtaining and managing seL4 project sources is through the use of Google's Repo tool.
```

lsf37 commented on Tutorials/seL4Kernel/setting-up.md:
````diff
@@ -0,0 +1,237 @@
+  sudo apt-get install repo
+```
+
+<details markdown='1'>
+<summary style="display:list-item">More on Repo</summary>
````
Those details sections are nice!

lsf37 commented on Tutorials/seL4Kernel/setting-up.md:
````diff
@@ -0,0 +1,237 @@
+(Some distributions use `pip` for python3, others use `pip3`. Use the Python 3 version for your distribution.)
+
+
+## Getting a build environment
+To get a running build environment for seL4 and Camkes, run:
+
+```bash
+git clone https://github.com/seL4/seL4-CAmkES-L4v-dockerfiles.git
+cd seL4-CAmkES-L4v-dockerfiles
+make user
````
These are now docker instructions again. They should be in the docker section.

lsf37 commented on Tutorials/seL4Kernel/setting-up.md:
````diff
@@ -0,0 +1,237 @@
+
+Replace `/<path>/<to>/` to match where you cloned the git repo of the docker files.
+
+*Reminder:* Include the absolute path to your `seL4-CAmkES-L4v-dockerfiles` folder, e.g.
+
+```bash
+echo $'alias container=\'make -C //home/jblogs/seL4-CAmkES-L4v-dockerfiles user HOST_DIR=$(pwd)\'' >> ~/.bashrc
````
`~/` instead of `/home/jblogs/` should work fine

lsf37 commented on Tutorials/seL4Kernel/setting-up.md:
```diff
@@ -0,0 +1,237 @@
+
+This then allows you to run `container` from any directory.
+
+*Reminder:* Restart Ubuntu or run a new terminal for the changes in `.bashrc` to take effect.
```
```suggestion
*Reminder:* start a new terminal for the changes in `.bashrc` to take effect or run `source ~/.bashrc`.
```

lsf37 commented on Tutorials/seL4Kernel/setting-up.md:
```diff
@@ -0,0 +1,237 @@
+
+A good workflow is to run two terminals:
+
+ - Terminal A is just a normal terminal, and is used for git operations, editing (e.g., vim, emacs), and other normal operations.
```
```suggestion
 - Terminal A is just a normal terminal, and is used for git operations, editing (e.g., vim, emacs, vscode), and other normal operations.
```

lsf37 commented:
(have rebased and removed the edits to `Gemfile.lock` + `.ruby-version`)

bbrcknl commented:
> * The new inline solutions and the resulting how-to guide are really cool!

Thanks!

> * I still need to go through all the parts, but I also like the new flow. There are some things to update in the details, though:
>   
>   * the "about seL4 page" in the beginning doesn't really fit. It's too long-winded and not aimed at tutorial level. We should probably start with the overview page instead (also not the Microkit).

I've removed the "about seL4" and started with "overview" instead

>   * the Microkit page calls the Microkit "an operating system framework". I don't think that name is right, and I'm not sure the page as such helps much. We might want to replace links to this page directly with links to the Microkit tutorial (not sure if that always works).

I got the wording from https://docs.sel4.systems/projects/microkit/

In any case, I've removed the GettingStarted page that was dedicated to the Microkit tutorial

>   * the overview page says "a basic understanding of seL4, e.g. by having done the Microkit tutorial." -- the Microkit tutorial is mostly hiding seL4 and not really providing much understanding on what seL4 is or does. The seL4 white paper might be a better resource for this one. I'm fine with pointing people at the Microkit tutorial as a potential very first step, but it needs to be clear that you won't really learn much about seL4 that way -- you'll learn how to build systems with a static architecture on top of seL4 there, which might be all that people want, but it's slightly different to the rest.

In the new "pathways" page (which lists alternative pathways through the tutorials), I have suggested that new developers start with Microkit as a very first step, with the caveat that Microkit hides seL4 mechanisms. Can take it out if you think it's misleading.

>   * the overview page doesn't really give an overview of the tutorials. It should. It should explain when you want to do/use Microkit, seL4 directly and/or CAmkES. Also what MCS is, what the dynamic libraries are about, and that there is Rust support.

Details have now been added.

>   * the prerequisites look a bit scary -- it's good that they are there, but it looks like I first have to read all of these before I can run "Hello world", which is not the case. You do need a background in OS for many of the other things, so I don't think we should remove the section, but we might for instance kick out the two wikipedia links that aren't really much good for learning anything (and if you do read the OS book, you'll get enough background for these).

I removed the wikipedia links and changed "prerequisites" to (the hopefully less scary) "recommended reading"

> * the new front page doesn't really work for me -- possibly because I have the browser on half screen size by default (which is similar to tablet size), and having 3 columns there looks not so great. The old 2 column design is about as much as can fit with that column size. How about leaving the main design intact for now, just changing the links under "Tutorials" there? (and making "Tutorials" itself not a link). Could still be tutorials, howtos, and manual. I would possibly add API Reference if there is space

Went back to the two-column format and put howtos and manual under "Tutorials"

> * a lot of the general menu links in unrelated parts of the site are now rendered as going off-site (probably just some small tweak somewhere, but needs to be fixed)

Can you pls give an example? Do you mean links that go from the docsite to the seL4 site, e.g. https://sel4.systems/Info/Docs/seL4-manual-latest.pdf?

> * I'd replace the links to forum, etc in the Resources section of the nav section on the left with a single link to a contact page

Done. I added a "contacts" section to the "Resources" front page, because this page already included links to mailing lists. "Contacts" now has links to mailing lists and forums.

> * Not sure the rename from "Getting Started" into "Working with seL4" is a success. The problem is more the content of that page, not so much what it is called ;-). In any case, "Working with seL4" is too long -- it line-breaks on narrower screens and looks bad when it does. Whatever name we pick should be narrower than the min size of the rest of the card on the front page. It'd be great to have an actual good Getting Started page, but that might lead a bit far astray from the purpose of this PR. I'd be fine with removing the links to that page from the top-level navigation and calling the heading on the front page "Resources" or something generic like that (without link to this specific page).

It's been renamed to "Resources." This page had an index page associated with it already. It's a bit verbose, but I mostly left it alone since it wasn't part of this PR. (I added a "forums" section, as described above.)

bbrcknl commented:
Renaming the GettingStarted directory to Resources will break a number of links pointing to the docsite.

Repos that will need updating:

- sel4/camkes-vm-examples-manifest
- sel4/seL4
- sel4/seL4_tools
- sel4/sel4bench-manifest
- sel4/sel4test-manifest
- sel4/sel4test

While the renaming causes some pain, it may be worth doing since GettingStarted is an inaccurate name. (However it may be outside the scope of this PR.)

bbrcknl commented:
@lsf37 

> a lot of the general menu links in unrelated parts of the site are now rendered as going off-site (probably just some small tweak somewhere, but needs to be fixed) 

I changed template.py in https://github.com/seL4/sel4-tutorials/pull/106 so that docsite links don't render as going offsite.

bbrcknl commented on Tutorials/capabilities.md:
```diff
@@ -4,7 +4,11 @@ title: Capabilities
+layout: tutorial
 SPDX-License-Identifier: CC-BY-SA-4.0
-SPDX-FileCopyrightText: 2020 seL4 Project a Series of LF Projects, LLC.
+SPDX-FileCopyrightText: 2024 seL4 Project a Series of LF Projects, LLC.
```
I'll put a note in my calendar to make the change in 2070 :-D

bbrcknl commented on Tutorials/dynamic-1.md:
```diff
@@ -3,9 +3,13 @@ toc: true
 ---
 {% include tutorial.md %}
 
+
+
```
Nope. D'oh.

bbrcknl commented on _includes/nav-sidebar.html:
```diff
@@ -11,7 +11,7 @@
   {% for entry in item.subfolderitems %}
         <li class="{% if page.url == entry.url %}active{% endif %}">
-          <a class="" href="{{ entry.url }}">
+          <a class="" href="../{{ entry.url }}">
```
I honestly can't remember how that crept into the updated code. I've changed it back.

bbrcknl commented on _includes/nav-sidebar.html:
```diff
@@ -48,21 +48,71 @@
     {% assign url = "/projects/" | append: new_project[0].name | append: '/' %}
         <li class="{% if page_url[2] == new_project[0].name %}active{% endif %}">
-            <a class="" href="{{url}}">{{new_project[0].display_name}}</a>
+            <a class="" href="../{{url}}/tutorials/">{{new_project[0].display_name}}</a>
```
Again, reverted to original code.

bbrcknl commented on _includes/nav-sidebar.html:
```diff
@@ -48,21 +48,71 @@
-            <a class="" href="{{t.url}}">{{t.title}}</a>
-        <li>
-  {% endfor %}
+{% if page_url[1] == "tutorials" %}
+  {% assign url = "" %}
```
The line
```
{% if page_url[1] == "tutorials" %}
```
is used to only capture tutorial pages, and to use the code below to generate a different navbar for the tutorial pages compared to the rest of the docsite.
I'm sure there's a better way of doing this, but I don't know how best to do it.

bbrcknl commented on _includes/nav-sidebar.html:
```diff
@@ -48,21 +48,71 @@
-
-{% if page_url[1] == "Tutorials" %}
-    <ul class="nav nav-sidebar">
-  {% assign tutorials = site.pages | where_exp: 'page', 'page.tutorial' | sort: 'tutorial-order' %}
-  {% for t in tutorials %}
-        <li class="{% if page.url == t.url %}active{% endif %}">
-            <a class="" href="{{t.url}}">{{t.title}}</a>
-        <li>
-  {% endfor %}
```
I will need your help with this. The code deleted above was for ordering the tutorials on the old (current) docsite, which was pretty shoddy. I.e. the tutorial ordering made little sense, especially to the newbie.

Indanz replied:
The Jekyll way of doing this would be to make the tutorials a collection, add an ordering property to the YAML front matter and add a [`sort_by`](https://jekyllrb.com/docs/collections/#sort-by-front-matter-key) to the config file to order it by that.

But this only works if all items are files in the repo. If that's not the case, things get more messy and the best solution is less clear. Another downside is that there is no overview of the order and you need to see the rendered result.

For the website I added [menu.json](https://github.com/seL4/website/blob/master/_data/menu.json) to have all metadata and ordering information in one place and used that as input to [generate](https://github.com/seL4/website/blob/master/_includes/menu.html) the wanted structure. The main reason I did it that way is to keep all the data compact with submenu support and in one place. It's also a mix of local and remote urls, so I couldn't add the metadata to front matter and I really didn't want to create shim files just for that as that's just messy.

Frankly, docs is a mess: It has all the tutorial shim files in the Tutorials directory which already contain a `tutorial-order` property, but it also added `_data/projects/sel4-tutorials.yml`. You only need one place to keep metadata.

What was wrong with the old code other than the order? The order can be changed by changing `tutorial-order`, it's probably alphabetically sorted by value. Main thing missing is the grouping you added, but that can be solved by adding a new field `category` to the first tutorial of each group: If set, generate a new '`<li>`'.

Once you know the order, it should also be possible to automatically generate the `Next tutorial` links too. Maintaining that manually will be hell and easy to forget when changing the order or adding new tutorials in the future.

You want to keep metadata out of your template code and you don't want the same metadata in multiple places.

lsf37 replied:
There are additional menu items in the tutorial side bar that are a mix of external links and other pages on the site. It also has non-link section headers. The nicest way would probably be similar to what you did for the main site, with one separate json file that has the structure and a separate `tutorials-nav` include.

lsf37 replied:
Hm. Nothing is ever simple. The `sel4-tutorials.yml` is currently there to define the "project", which has info that gets rendered on a bunch of overview pages, so we can't just remove it. We *can* use it to define the order, because it has an ordered list of all tutorial pages, and we can add additional info to each of the entries, but I'm not sure how to add additional info like section headers, external links, and links to other pages.

I guess, we could add another field "nav-section" to each component in there and then another top-level entry with a list of section headers, where each either has a nav section name or directly a list of nav items consisting of text and link. That could work. Not entirely sure I'll have time this afternoon, but I could give this a go.

Indanz replied:
I would go for the minimal change: Keep the non-tutorial links hard coded here for now and use the existing `tutorial-order` to generate the tutorial part, with a new property to create submenus. And then you can use that to figure out the next tutorial link too.

lsf37 replied:
I couldn't get that quite working -- a lot of the hard coded part also needs to figure out "next" links, which is then not available as data. I went with something similar to the above, which removes a bit of the duplication (order and description in front matter) with sel4-tutorials.yml. The side bar works out ok, computing a reasonable "next" link did not end up to be very elegant.

Choosing the dict keys slightly differently would remove some of the case distinctions, but the ones prescribed by `sel4-tutorials.yml` sounded too nonsensical in the nav bar (`name` refers to file name, which I found most confusing in there. I guess `display_name` would be usable). The names in `sel4-tutorials.yml` is prescribed by other generic code on the site, so we can't tweak that without larger changes.

Anyway, have a look if the current version is acceptable. If not, we can always revert some of this.

I have not yet removed the manual "next" links for comparison. One thing that this brought up is that we have two "Overview" pages, which we should probably avoid.

lsf37 commented on Hardware/IA32.md:
```diff
@@ -29,7 +29,7 @@ menuentry "Load seL4 VM" --class os {
     insmod ext2
     set root='(hd0,msdos2)'
-    multiboot /boot/sel4kernel
+    multiboot /boot/seL4
```
this one looks like it got caught in a global search/replace. It should stay on `sel4kernel`

lsf37 commented on Hardware/IA32.md:
```diff
@@ -59,13 +59,13 @@ install-mbr /dev/sdb
 cp images/sel4test-driver-image-ia32-pc99 /mnt/rootserver
-cp images/kernel-ia32-pc99 /mnt/sel4kernel
+cp images/kernel-ia32-pc99 /mnt/seL4
 cat > /mnt/syslinux.cfg <<EOF
 SERIAL 0 115200
 DEFAULT seL4test
 LABEL seL4test
     kernel mboot.c32
-    append sel4kernel --- rootserver
+    append seL4 --- rootserver
```
these two as well

bbrcknl commented on _includes/nav-sidebar.html:
```diff
@@ -48,21 +48,71 @@
+    <ul>
+      <li><a href="{{url}}/tutorials/mcs">MCS Extensions</a></li>
+    </ul>
+    <li>Dynamic libraries</li>
```
I've changed it to "Libraries" here and in the tutes PR
https://github.com/seL4/sel4-tutorials/pull/106

bbrcknl replied:
This has broken a bunch of things, which I'll fix on Friday.

lsf37 commented on _layouts/tutorial.html:
```diff
@@ -0,0 +1,32 @@
+    {% include toc-sidebar.html %}
+  {% endif %}
+</div>
+<script src="{{ base.url | prepend: site.url }}/assets/js/toggle-markdown.js"></script>
```
This should be slightly more stable when we change jekyll versions:
```suggestion
<script src="{{ '/assets/js/toggle-markdown.js' | relative_url }}"></script>
```
(`relative_url` produces an absolute path from the site root taking into account any base url and other config settings Jekyll is applying)

lsf37 commented on _includes/header.html:
```diff
@@ -16,10 +16,10 @@
         <h2><a href="/processes">Contributing</a></h2>
         <h2><a href="/projects">Projects</a></h2>
-        <h2><a href="/Tutorials">Tutorials</a></h2>
+        <h2><a href="/tutorials/">Tutorials</a></h2>
```
So this change doesn't appear for me on the file system -- the directory there is called `Tutorials` with capital `T` for me, and the links to lower-case `tutorials` are not working because of that.

Which one is the intended one? (The Mac filesystem is case insensitive, so my repo checkout would not have seen a rename from the `Tutorials` directory to `tutorials`, it'd have stayed the same).

lsf37 replied:
Looking into it further, it gets introduced in [this commit](https://github.com/seL4/docs/commit/68c00ea5893404402807c44fe599f4e2d5334aae) which might mean it was an attempt to get the link rendering working correctly. The rename `Tutorials` -> `tutorials` should be reverted, though.

lsf37 commented on Tutorials/get-the-tutorials.md:
````diff
@@ -0,0 +1,32 @@
+## Python Dependencies
+Additional python and CAmkES dependencies are required to build [tutorials](ReworkedTutorials). To install you can run:
+```
+pip3 install --user aenum pyelftools camkes-deps
````
Only `camkes-deps` is required, `aenum` and `pyelftools` are part of `camkes-deps` already.

```suggestion
pip3 install --user aenum pyelftools camkes-deps
```

lsf37 commented on Tutorials/get-the-tutorials.md:
```diff
@@ -0,0 +1,32 @@
+
+# Getting the tutorials
+## Python Dependencies
+Additional python and CAmkES dependencies are required to build [tutorials](ReworkedTutorials). To install you can run:
```
```suggestion
The CAmkES python dependencies are required to build the [tutorials](ReworkedTutorials). To install you can run:
```

lsf37 commented on Tutorials/get-the-tutorials.md:
````diff
@@ -0,0 +1,32 @@
+```
+pip3 install --user aenum pyelftools camkes-deps
+```
+*Hint:* This step only needs to be done once, i.e. before doing your first tutorial
````
```suggestion
*Hint:* This step only needs to be done once, i.e. before doing your first tutorial.
```

Indanz reviewed (commented):
You accidentally committed a conflicted merge, see ``<<<<<<<``, `>>>>>>>` blocks.

lsf37 commented:
> You accidentally committed a conflicted merge, see `<<<<<<<`, `>>>>>>>` blocks.

Sorry about that, should now be fixed. Not sure how I managed to miss that one.

lsf37 commented on content_collections/_updates/sel4-tutorials/camkes-3.8.x.md:
```diff
@@ -16,18 +16,18 @@ SPDX-FileCopyrightText: 2020 seL4 Project a Series of LF Projects, LLC.
 - `hello-camkes-2`: Update exit text test to match actual output.
 - `mcs`: reduce spinner budget for final task to ensure timeout behavior happens correctly.
-- `dynamic-4`: Correctly initalize a stack variable.
+- `libraries-4`: Correctly initalize a stack variable.
```
In these older release notes we shouldn't change the name -- in the version it refers to, they were still called `dynamic-`.

Indanz commented on Makefile:
```diff
@@ -51,7 +51,7 @@ _repos/tutes:
 	PYTHONPATH=_repos/sel4/capdl/python-capdl-tool _repos/sel4proj/sel4-tutorials/template.py --docsite --out-dir _repos/tutes --tut-file $</$(@F)
 
-TUTORIALS:= $(filter-out index.md,$(notdir $(wildcard Tutorials/*.md)))
+TUTORIALS:= $(filter-out index.md get-the-tutorials.md how-to.md pathways.md setting-up.md,$(notdir $(wildcard Tutorials/*.md)))
```
Maybe add a comment why these are filtered out.

Indanz commented on Tutorials/hello-world.md:
```diff
@@ -2,9 +2,9 @@
+layout: tutorial
 SPDX-License-Identifier: CC-BY-SA-4.0
-SPDX-FileCopyrightText: 2020 seL4 Project a Series of LF Projects, LLC.
+SPDX-FileCopyrightText: 2024 seL4 Project a Series of LF Projects, LLC.
```
Please keep all copyright years as they were originally, or if you want to add to it, at least change it to 2020-2024.

Indanz commented on Tutorials/index.md:
```diff
@@ -1,27 +1,30 @@
+layout: tutorial
 SPDX-License-Identifier: CC-BY-SA-4.0
-SPDX-FileCopyrightText: 2020 seL4 Project a Series of LF Projects, LLC.
+SPDX-FileCopyrightText: 2024 seL4 Project a Series of LF Projects, LLC.
```
Idem.

Indanz commented on _includes/header.html:
```diff
@@ -16,10 +16,10 @@
       <nav aria-label="Banner links">
-        <h2><a href="/GettingStarted">Getting Started</a></h2>
+        <h2><a href="/Resources">Resources</a></h2>
         <h2><a href="/processes">Contributing</a></h2>
         <h2><a href="/projects">Projects</a></h2>
-        <h2><a href="/Tutorials">Tutorials</a></h2>
+        <h2><a href="/Tutorials/">Tutorials</a></h2>
```
I wonder, should these go through ` | relative_url ` to work properly in the Github preview?

bbrcknl replied:
Like this?

```
        <h2><a href="{{ relative_url }}/Resources" />Resources</h2>
        <h2><a href="{{ relative_url }}/processes" />Contributing</a></h2>
        <h2><a href="{{ relative_url }}/Projects" />Projects</h2>
        <h2><a href="{{ relative_url }}/Tutorials" />Tutorials</h2>
```

lsf37 replied:
The syntax goes like this:
```
<h2><a href="{{ '/Resources' | relative_url }}" />Resources</h2>
```

bbrcknl replied:
Thanks!

Indanz commented on assets/js/toggle-markdown.js:
```diff
@@ -0,0 +1,15 @@
+    document.body.querySelectorAll('details')
+      .forEach((e) => {(e.hasAttribute('open')) ?
+        e.removeAttribute('open') : e.setAttribute('open',true);
+        console.log(e.hasAttribute('open'))
```
This one is still unresolved.

Indanz reviewed (approved):
Other than some nitpicks, all looks good.

bbrcknl commented:
> Other than some nitpicks, all looks good.

Great, thanks for your review. I believe that all the outstanding issues have now been resolved.

Indanz commented on Tutorials/how-to.md:
```diff
@@ -9,71 +9,70 @@ This guide provides links to tutorial solutions as quick references for seL4 cal
  - [Calculate the size of a CSpace](capabilities.md#how-big-is-your-cspace)
  - [Copy a capability between CSlots](capabilities.md#copy-a-capability-between-cslots)
  - [Delete a capability](capabilities.md#how-do-you-delete-capabilities)
  - [Suspend a thread](capabilities.md#suspend-a-thread)
```
I'm afraid you need to add `?tut_expand` to all the links for it to work, now only the top level links expand the hints.

Maybe using a variable is nicer than all the repetitions of `capabilities?tut_expand`.

bbrcknl replied:
It actually does work with just the top level links. The `js` code looks for everything that's a `details` markdown.

Indanz replied:
Yes, everything gets expanded if you follow the top link. But if you follow one of the sublinks, `tut_expand` won't be part of the URL and that code does nothing.

bbrcknl replied:
Blast, ok. I'll fix it on Wed.

Ivan-Velickovic commented:
I get this when I try to build:
```
make[1]: *** No rule to make target '_repos/tutes/libraries-1.md', needed by 'tutorials'.  Stop.
make[1]: Leaving directory '/docs'
```

lsf37 commented:
> I get this when I try to build:
> 
> ```
> make[1]: *** No rule to make target '_repos/tutes/libraries-1.md', needed by 'tutorials'.  Stop.
> make[1]: Leaving directory '/docs'
> ```

You need to manually set the right branch in the tutorials repo after it has pulled it, but before it builds it. It's unpleasant, but we're almost there.

lsf37 commented:
The docsite should really be using a manifest and not just randomly clone things, but one thing at a time.

Ivan-Velickovic commented:
Ah yes, thanks.

lsf37 reviewed (approved):
Now squashed into one commit and ready to go.
